File: fims_tmb_exercise/randomwalk.R
setwd("~/Documents/Github/homework/fims_tmb_exercise")

library(TMB)
compile("randomwalk.cpp")
dyn.load(dynlib("randomwalk"))

## Read data
mu <- 0.75
sigma_proc <- 1
sigma_obs <- 1

sim_num <- 200
set.seed(999)
eta <- u <- rep(NA, sim_num)
u[1] <- rnorm(n = 1, mean = 0, sd = sigma_proc)
for (i in 2:sim_num) {
  eta[i] <- mu + u[i - 1]
  u[i] <- rnorm(1, eta[i], sigma_proc)
}
y <- rnorm(sim_num, u, sigma_obs)
data <- list(y = y)

## Parameter initial guess
parameters <- list(
  u = y,
  sigma_proc = 1,
  sigma_obs = 1,
  mu = 0.7
)

## Fit model
obj <- MakeADFun(data, parameters, random = "u", DLL = "randomwalk")
newtonOption(obj, smartsearch = FALSE)
obj$fn()
obj$gr()
system.time(opt <- nlminb(obj$par, obj$fn, obj$gr))
rep <- sdreport(obj)
rep

## Simulation
set.seed(1)
obj$simulate()

plot(obj$simulate()$y, y,
  xlab = "TMB simulated y",
  ylab = "R simulated y"
)

## Validation
u <- c(0, cumsum(rnorm(sim_num - 1, mean = mu, sd = sigma_proc)))
y <- u + rnorm(sim_num, sd = sigma_obs)

data <- list(y = y)
parameters <- list(u = u, sigma_proc = sigma_proc, sigma_obs = sigma_obs, mu = 0)

obj <- MakeADFun(data, parameters, random = c("u"), DLL = "randomwalk", map = list(mu = factor(NA)))
opt <- do.call("optim", obj)
sdr <- sdreport(obj)
estu <- summary(sdr, "random")

plot(u, type = "l", xlab = "Time", ylab = "u", lwd = 3)
points(y)
lines(estu[, 1], lwd = 1, col = "red")

resid <- y - estu[, 1]
Norm.resid <- resid / estu[, 2]
acf <- acf(Norm.resid, plot = FALSE)
predict <- oneStepPredict(obj, observation.name = "y", data.term.indicator = "keep", method = "oneStepGeneric")
qqnorm(predict$residual)
abline(0, 1)
ks.test(predict$residual, "pnorm")

File: fims_tmb_exercise/randomwalk.cpp
// A randomwalk model for FIMS TMB exercises
#include <TMB.hpp>

template<class Type>
  Type objective_function<Type>::operator() ()
{
  /* Data section */
  DATA_VECTOR(y);
  DATA_VECTOR_INDICATOR(keep, y);  // For one-step predictions
  
  /* Parameter section */
  PARAMETER_VECTOR(u);
  PARAMETER(sigma_proc);
  PARAMETER(sigma_obs);
  PARAMETER(mu);
  
  /* Procedure section */
  int timeSteps=y.size();
  
  vector<Type> eta(timeSteps);
  
  for(int i=1;i<timeSteps;i++){
    eta[i] = mu + u[i-1];
  }
  
  Type nll = 0.0;
  
  for(int i=1;i<timeSteps;i++){
    nll -= dnorm(u[i], eta[i], sigma_proc, 1);
  }
  
  for(int i=0;i<timeSteps;i++){
    // nll -= dnorm(y[i], u[i], sigma_obs, 1);
    nll -= keep[i] * dnorm(y[i], u[i], sigma_obs, true);
  }
  
  SIMULATE {
    y = rnorm(u, sigma_obs);  // Simulate response
    REPORT(y);          // Report the simulation
  }
  
  REPORT(eta);
  ADREPORT(eta);

  REPORT(u);
  ADREPORT(u);

  REPORT(sigma_obs);
  ADREPORT(sigma_obs);

  REPORT(sigma_proc);
  ADREPORT(sigma_proc);
  
  return nll;
}
